crypto/bio/bio.c

```
@@ -68,34 +68,115 @@
-                                        const void *buf, int len, int ret) {
-  if (HAS_CALLBACK(bio)) {
-    size_t processed = 0;
-    // The original BIO return value can be an error value (less than 0) or
-    // the number of bytes read/written
-    if (ret > 0) {
-      processed = ret;
+// |callback_fn_wrap_ex| adapts the legacy callback interface |BIO_callback_fn| to the
+// extended callback interface |BIO_callback_fn_ex|. This function should only be
+// called when |callback_ex| is not available and the legacy callback is set.
+//
+// The extended interface parameters |len| and |processed| are mapped to the legacy
+// interface parameters |argi| and |bio_ret| respectively.
+//
+// Returns -1 on NULL |BIO| or callback, otherwise returns the result of the legacy
+// callback.
+static long callback_fn_wrap_ex(BIO *bio, int oper, const char *argp,
+                              size_t len, int argi, long argl, int bio_ret,
+                              size_t *processed) {
+  assert(bio != NULL);
+  assert(bio->callback != NULL);
+
+  /* Strip off any BIO_CB_RETURN flag */
+  int bareoper = oper & ~BIO_CB_RETURN;
+
+  if (bareoper == BIO_CB_READ || bareoper == BIO_CB_WRITE
+    || bareoper == BIO_CB_GETS) {
+    /* In this case |len| is set, and should be used instead of |argi| */
+    if (len > INT_MAX) {
+      return -1;
     }
-    // Pass the original BIO's return value to the callback. If the callback
-    // is successful return processed from the callback, if the callback is
-    // not successful return the callback's return value.
-    long callback_ret = bio->callback_ex(bio, oper, buf, len, 0, 0L, ret, &processed);
-    if (callback_ret <= INT_MAX && callback_ret >= INT_MIN) {
-      ret = (int)callback_ret;
-      if (ret > 0) {
-        // BIO will only read int |len| bytes so this is a safe cast
-        ret = (int)processed;
-      }
+
+    argi = (int)len;
+  }
+
+  if (bio_ret && (oper & BIO_CB_RETURN) && bareoper != BIO_CB_CTRL) {
```

Should this be bio_ret or bio_ret > 0? Could bio_ret be a negative value? What should happen then?

---

`bio_ret `having a negative value would allow `BIO_CB_RETURN ` operations not including `BIO_CB_CTRL `to set `processed `to `bio_ret`.

You are correct, currently this could overwrite error values of `bio_ret `with `processed ` masking the original error condition.